### DISCUSSION TITLE
How do I/Can I insert a Quarto variable into an executable code block?

### DISCUSSION THREAD

**Discussion by jreades** [quarto-dev/quarto-cli#9326] [Q&A]
### Description

I am trying to parameterise the source data path in my course's practicals since the data changes every year. The basic idea is to update the one parameter and then re-render the entire project such that the outputs are what my students will get when they follow the practical sessions.

So I want to have a Quarto variable or metadata entry that is the current year's data path. This works fine in *non*-executing code blocks, but not in code blocks that are going to be evaluated during the rendering process. I found in the documentation that this is a feature, not a bug, since they are *supposed* to be evaluated later in the rendering pipeline.

So what *is* the right/preferred way to achieve this? An environment variable? Imported `.py` file? Some other mechanism?

For a skeletal MVP here's what my current codebase looks like:

### YAML Conf

In the `_variable.yml` or `_metadata.yml` file:

```yaml
data:
  fn: "<some_file_name>"
```

### Question

````qmd

```python
# Set download URL
host = '<some_host_name>'
path = '~<some_user_name>/{{< var data.fn >}}'
url  = f'{host}/{path}'

# your code here
df = pd.read_csv(??, compression='gzip', low_memory=False)
print(f"Data frame is {df.shape[0]:,} x {df.shape[1]}")
```

````

### Answer

````qmd

```{python}
# Notice that we import a specific error here
# but other types of errors (e.g. NameError)
# are always available because they're so basic.
from urllib.error import URLError

# Set download URL -- you can also find reviews and calendar
# at the same location (i.e. {{< var data.year >}}-{{< var data.month >}}-{{< var data.day >}}-calendar.csv.gz)
host = '<some_host_name>'
path = '~<some_user_name>/{{< var data.fn >}}'
url  = f'{host}/{path}'

if not os.path.exists('tmp.csv.gz'):
    try:
        df = pd.read_csv(url, compression='gzip', low_memory=False)
        df.to_csv('tmp.csv.gz', compression='gzip', index=False)
    except URLError as e:
        print(f"Unable to access: {url}")
        print(e)
else:
    df = pd.read_csv('tmp.csv.gz', compression='gzip', low_memory=False)

try:
    print(f"Data frame is {df.shape[0]:,} x {df.shape[1]}")
except NameError as e:
    print("Unable to load the data")

```

````

So the 'Question' is interpolated as I'd like, but the 'Answer' obviously just displays the raw QMD `var` syntax. I assume I'm missing something but can't seem to find any examples of a recommended way to achieve this.

Many thanks!

**Comment by mcanouil** [comment]
Could you properly format your post using code blocks for code and terminal outputs? Thanks.
If your code contains code blocks, you need to enclose it using more backticks, *i.e.*, usually four ` ```` `.

`var` shortcode is not the same thing as parameters, see <https://quarto.org/docs/computations/parameters.html>.

**Reply by jreades** [reply]
Sorry, have sorted that out now -- I couldn't figure out how to get back to the documentation for formatting questions once I'd started writing the question!

**Comment by jreades** [comment]
UPDATE: Based on initial feedback it looks like parameterisation at runtime/render time is handled differently via command line [Parameters](https://quarto.org/docs/computations/parameters.html).

So the approach seems to be:

```bash
quarto render practicals/Practical-05-Numeric_Data.qmd --execute-params _params.yml
```

Where the YAML file might be:
```yml
data:
  fn: "2023-09-06-listings.csv.gz"
  year: "2023"
  month: "09"
  day: "06"
```

However, this has now led me to:

```python
ERROR:
The papermill package is required for processing --execute-params
```

As far as I can tell papermill is (now) installed since `python -c 'import papermill'` does _not_ return an error.

**Reply by mcanouil** [reply]
In case you use environment such as Virtual environment or Conda, it needs to be installed in those.
Quarto will use the activated environment and what is inside it.

**Reply by jreades** [reply]
Yes, that's what I figured -- I'm currently doing everything in the `base` environment for simplicity:

```bash
conda activate base
conda install -c conda-forge papermill
quarto render practicals/Practical-05-Numeric_Data.qmd --execute-params _params.yml
```

But that still returns a missing papermill install error. I'll keep at it.

**Reply by jreades** [reply]
Gah, new terminal window and I'm sorted so must have been an env issue. SORRY!

**Reply by jreades** [reply]
Right, so I'm now at this point:

````qmd
```{python}
#| tags: [parameters]
#| echo: false

year  = "2023"
month = "09"
day   = "06"
fn    = f"{year}-{month}-{day}-listings.csv.gz"
```
````

And my QMD file now looks like this which, for reasons that should be fairly obvious, isn't working:

````qmd

:::: {.panel-tabset}

#### Question

```python
# Set download URL
host = '<some host name>'
path = '<some path>/`{python} f"{fn}"`'
url  = f'{host}/{path}'

# your code here
df = pd.read_csv(??, compression='gzip', low_memory=False)
print(f"Data frame is {df.shape[0]:,} x {df.shape[1]}")
```

#### Answer

```{python}
# Notice that we import a specific error here
# but other types of errors (e.g. NameError)
# are always available because they're so basic.
from urllib.error import URLError

# Set download URL -- you can also find reviews and calendar
# at the same location (i.e. `{python} f"{year}-{month}-{day}-calendar.csv.gz"`)
host = '<some host name>'
path = f'<some path>/{fn}'
url  = f'{host}/{path}'

if not os.path.exists('tmp.csv.gz'):
    try:
        df = pd.read_csv(url, compression='gzip', low_memory=False)
        df.to_csv('tmp.csv.gz', compression='gzip', index=False)
    except URLError as e:
        print(f"Unable to access: {url}")
        print(e)
else:
    df = pd.read_csv('tmp.csv.gz', compression='gzip', low_memory=False)

try:
    print(f"Data frame is {df.shape[0]:,} x {df.shape[1]}")
except NameError as e:
    print("Unable to load the data")
```

::::

````

The *first* python block in the **Question** section (which isn't evaluated at render-time) gives me a `fn` parameter that has the hyphens escaped: `path = '<some path>/2023\-09\-06\-listings\.csv\.gz'`

The *second* python block in the **Answer** section (which *should* be evaluated so that an answer is presented) doesn't do the interpolation that I'd been hoping to see (i.e. the `path` would should the actual path to the data taken from the parameter.

Hope this makes sense. To try to be clear; I want some parameters/variables/meta (I don't particularly care which) to be interpolated into the code block *prior* to the engine being run on that code block.

**Comment by jreades** [comment]
OK, so now I'm looking at pre- and post-render scripts, which also seems *nearly* meet my requirements but I'm finding the documentation a bit tricky to line up with my ideal conceptualisation of how this process could work. 

My working idea is now that a pre-render script could match on some pattern (e.g. `{<< (...) >>}`) and interpolate the parameter or environment variable using the matched name (e.g. `re.sub('{<< (\S+) >>}', lambda m: os.environ[m.group()], <string>)`) *before* evaluation of any code. So this should get me around the different ways the code blocks are/are not executed and I just treat the QMD as a text file.

The issue I'm having, conceptually, is that it's unclear to me *where* this work should be/could be done: I don't want to modify the original source QMD files (losing the `{<< ... >>}` markers), and I'm assuming that only fully-rendered (post-code execution) documents are put in the `output` folder. I printed out the `cwd` for the pre-render script and it just shows up as the document root.

Is the only way to do this to write each file to, say, a `tmp` file and then tell Quarto to render *that* file instead of the original QMD? This feels really clunky for something that, intuitively, should be relatively 'easy' but clearly is either an edge case or incompatible with Quarto's pipeline model.

**Reply by mcanouil** [reply]
Your example makes it way more complicated to understand and help than it needs to be.
Could you make it simple and complete? Also, even without the use of shortcodes or parameters, your code cannot work, so please do share something simple and usable (You'll adapt after for your use case).

For example:

````qmd
---
title: "Quarto Playground"
format: html
data:
  msg: "Hello, Quarto!"
---

This is a playground for Quarto.

```{python}
#| tags: [parameters]
#| echo: false
dataMsg = "Hello, Quarto!"
```

```{python}
print("{{< meta data.msg >}}")
```

```{python}
print(f"{dataMsg}")
```

```{python}
print("{{< var data.msg >}}")
```

````

With `_quarto.yml` (empty) and `_variables.yml`:

```yaml
data:
  msg: "Hello, Quarto!"
```

<img width="418" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/83bab968-dc0c-4e28-a0b1-7485be714d5d">

From the command line, you can change the default for "meta" and "parameters" (still needs to be set in the document with some values):

```sh
--execute-params file.yml
--execute-param dataMsg:"Bye, Quarto!"
--metadata-files file.yml
--metadata data.msg:"Bye, Quarto!"
```

**Reply by jreades** [reply]
Sure, sorry to have confused things -- I was trying to provide context but have clearly provided too much. I believe that I 'get' all of your examples above but none of them *seem* to achieve what I'm trying to achieve.

## Simpler worked example

Here's something that demonstrates the behaviour I'm wrestling with and the output is attached as an image:

### Unrendered Code Blocks

So these are questions that will *not* be executed during the rendering process.

== Both of these will display exactly what I want the student to see in the *question* ==

````qmd
```python
path = "~/{{< var data.fn >}}"
print(path)
```
````

````qmd
```python
# The path is {{< var data.fn >}}
```
````

### Rendered Code Block

But these are answers that *will* be executed during the rendering process.

== Neither of these will show what I want the student to see in the *answer* since they still show the raw `var` ==

````qmd
```{python}
path = "~/{{< var data.fn >}}"
print(path)
```
````

````qmd
```{python}
# The path is {{< var data.fn >}}
```
````

In the first executed code example Quarto *does* successfully print the path (see image) but what the student still sees in their document is `path = '~/{{< var data.fn >}}'` and *not* `path = '~/2023-09-06-listings.csv.gz'`. In the same way, the comment does not appear to change in the output. I want these to *look the same* to the student because they are getting a prompt to write/complete/correct some code in their notebook and then should see the fully-rendered solution in the other panel.

<img width="822" alt="Screenshot 2024-04-11 at 11 21 11" src="https://github.com/quarto-dev/quarto-cli/assets/3224421/8ce8faf5-d116-4d99-abfa-ec08bcb22703">

To put it another way: why are the *Rendered* and *Unrendered* outputs *visibly* different?

**Reply by mcanouil** [reply]
I see, then you should use IPython and output markdown from your code cell.

edit:

Something like that

````qmd
---
title: "Quarto Playground"
format: html
data:
  msg: "Hello, Quarto!"
---

This is a playground for Quarto.


```{python}
#| echo: false
from IPython.display import Markdown

Markdown(
  '```{{python}}\n'
  'print("{{< meta data.msg >}}")\n'
  '```'
)
```

```{python}
#| echo: false
print("{{< meta data.msg >}}")
```
````

**Reply by jreades** [reply]
So close now, but in this solution the 'answer' no longer executes so the output is no longer generated. In my toy example the output is just a `print`, but obviously in my full use case this could be plots, tables, etc.

So in sketch form I would want something like:

````qmd
```{python}
#| echo: false
from IPython.display import Markdown
msg = """path = '~/{{< var data.fn >}}'
print(path)
"""
Markdown(msg)
eval(msg) # Doesn't work as is, but you get the sense
```
````

This would then show both the interpolated 'input' *and* the executed output. Hope this makes sense.

**Reply by mcanouil** [reply]
Unfortunately, in Python, there is not the "child" feature that `knitr` has.
Here you'll have to duplicate and have one code cell to execute and one to display.
See my example which renders as below

<img width="598" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/16575c27-076b-4955-a74c-769905b17083">
